File: CodingSample/DotNet/Fundamentals/Inheritance/InterfaceTypeTest/BankLib/Account.cs
namespace Banking;

//Only a type declared with 'public' keyword
//is visible outside of the current assembly/project.
//A class declared with 'abstract' keyword
//does not support activation.
public abstract class Account
{
    public long Id { get; internal set; }

    public decimal Balance { get; protected set; }    

    //A method declared using 'abstract' keyword
    //is pure - it does not provide any specific implementation
    public abstract void Deposit(decimal amount);

    public abstract void Withdraw(decimal amount);
}

File: CodingSample/Java/Core/Inheritance/InterfaceTest/bank/banking/Account.java
package banking;

//a class defined with 'abstract' keyword
//cannot be instantiated
public abstract class Account {
    
    long id;
    protected double balance;

    public long id() {
        return id;
    }

    public double balance() {
        return balance;
    }

    //a method defined with 'abstract' keyword
    //does mnot have a body
    public abstract void deposit(double amount);

    public abstract void withdraw(double amount) throws InsufficientFundsException;

    //a method declared with 'final' keyword 
    //cannot be overridden in the derived class
    public final void transfer(double amount, Account that) throws InsufficientFundsException {
        if(this == that)
            throw new IllegalTransferException();
        this.withdraw(amount);
        that.deposit(amount);
    }

}
